File: packages/capacitor-plugin/ios/Sources/UpdaterPlugin/BundleCrypto.swift
import CryptoKit
import Foundation
import os

enum BundleCryptoError: Error, LocalizedError {
  case noMatchingKey(String)
  case invalidParameter(String)
  case decryptionFailed(String)
  case insufficientMemory(encryptedBytes: Int64, budgetBytes: Int64)

  var errorDescription: String? {
    switch self {
    case let .noMatchingKey(kid):
      return "no matching bundle key (kid \(kid))"
    case let .invalidParameter(name):
      return "invalid bundle encryption parameter: \(name)"
    case let .decryptionFailed(detail):
      return "bundle decryption failed: \(detail)"
    case let .insufficientMemory(encryptedBytes, budgetBytes):
      return "insufficient_memory_for_encrypted_bundle; encryptedBytes=\(encryptedBytes); budgetBytes=\(budgetBytes)"
    }
  }
}

/// AES-256-GCM bundle decryption.
///
/// The CLI encrypts the zip with a random per-bundle DEK and wraps the DEK
/// under the app KEK; both ciphertexts carry the 16-byte GCM tag appended.
enum BundleCrypto {
  private static let tagLength = 16
  private static let keyLength = 32
  private static let reserveBytes: Int64 = 32 * 1024 * 1024
  private static let maximumEncryptedBytes: Int64 = 128 * 1024 * 1024

  static func requireMemoryBudget(encryptedBytes: Int64, availableBytes: Int64) throws {
    let budget = availableBytes <= reserveBytes ? 0 : min(maximumEncryptedBytes, (availableBytes - reserveBytes) / 4)
    guard encryptedBytes <= budget else {
      throw BundleCryptoError.insufficientMemory(encryptedBytes: encryptedBytes, budgetBytes: budget)
    }
  }

  static func unwrapDek(
    kek: Data,
    wrapNonceB64: String,
    wrappedDekB64: String
  ) throws -> Data {
    guard kek.count == keyLength else {
      throw BundleCryptoError.invalidParameter("bundle key length")
    }
    guard let wrapNonce = Data(base64Encoded: wrapNonceB64) else {
      throw BundleCryptoError.invalidParameter("wrapNonce")
    }
    guard let wrappedDek = Data(base64Encoded: wrappedDekB64),
          wrappedDek.count == keyLength + tagLength else {
      throw BundleCryptoError.invalidParameter("wrappedDek")
    }

    let dek: Data
    do {
      let sealedBox = try AES.GCM.SealedBox(
        nonce: AES.GCM.Nonce(data: wrapNonce),
        ciphertext: wrappedDek.prefix(keyLength),
        tag: wrappedDek.suffix(tagLength)
      )
      dek = try AES.GCM.open(sealedBox, using: SymmetricKey(data: kek))
    } catch {
      throw BundleCryptoError.decryptionFailed("DEK unwrap: \(error.localizedDescription)")
    }
    guard dek.count == keyLength else {
      throw BundleCryptoError.decryptionFailed("unwrapped DEK has unexpected length")
    }
    return dek
  }

  static func decryptFile(
    dek: Data,
    nonceB64: String,
    input: URL,
    output: URL,
    availableBytes: Int64 = Int64(os_proc_available_memory())
  ) throws {
    guard let nonce = Data(base64Encoded: nonceB64) else {
      throw BundleCryptoError.invalidParameter("nonce")
    }

    guard dek.count == keyLength, nonce.count == 12 else {
      throw BundleCryptoError.invalidParameter("key or nonce length")
    }
    let attributes = try FileManager.default.attributesOfItem(atPath: input.path)
    guard let size = attributes[.size] as? NSNumber, size.int64Value > tagLength else {
      throw BundleCryptoError.invalidParameter("ciphertext too short")
    }
    try requireMemoryBudget(encryptedBytes: size.int64Value, availableBytes: availableBytes)
    let ciphertextWithTag = try Data(contentsOf: input, options: .mappedIfSafe)
    guard ciphertextWithTag.count == size.intValue else {
      throw BundleCryptoError.invalidParameter("ciphertext changed during read")
    }

    do {
      let sealedBox = try AES.GCM.SealedBox(
        nonce: AES.GCM.Nonce(data: nonce),
        ciphertext: ciphertextWithTag.prefix(ciphertextWithTag.count - tagLength),
        tag: ciphertextWithTag.suffix(tagLength)
      )
      let plaintext = try AES.GCM.open(sealedBox, using: SymmetricKey(data: dek))
      try plaintext.write(to: output, options: .atomic)
    } catch let error as BundleCryptoError {
      throw error
    } catch {
      throw BundleCryptoError.decryptionFailed(error.localizedDescription)
    }
  }
}

File: packages/capacitor-plugin/android/src/main/java/com/otakit/updater/BundleCrypto.java
package com.otakit.updater;

import java.io.DataInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * AES-256-GCM bundle decryption.
 *
 * The CLI encrypts the zip with a random per-bundle DEK and wraps the DEK
 * under the app KEK; both ciphertexts carry the 16-byte GCM tag appended
 * (Java's AES/GCM/NoPadding expects exactly that layout).
 */
final class BundleCrypto {

  private static final int GCM_TAG_BITS = 128;
  private static final int KEY_LENGTH = 32;
  private static final int TAG_LENGTH = 16;
  private static final long RESERVE_BYTES = 32L * 1024 * 1024;
  private static final long MAX_ENCRYPTED_BYTES = 128L * 1024 * 1024;

  private BundleCrypto() {}

  static byte[] unwrapDek(byte[] kek, byte[] wrapNonce, byte[] wrappedDek) throws Exception {
    if (kek == null || kek.length != KEY_LENGTH) {
      throw new IllegalStateException("invalid bundle encryption parameter: bundle key length");
    }
    if (wrappedDek == null || wrappedDek.length != KEY_LENGTH + TAG_LENGTH) {
      throw new IllegalStateException("invalid bundle encryption parameter: wrappedDek");
    }
    Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
    cipher.init(
      Cipher.DECRYPT_MODE,
      new SecretKeySpec(kek, "AES"),
      new GCMParameterSpec(GCM_TAG_BITS, wrapNonce)
    );
    byte[] dek = cipher.doFinal(wrappedDek);
    if (dek.length != KEY_LENGTH) {
      throw new IllegalStateException("bundle decryption failed: unexpected DEK length");
    }
    return dek;
  }

  static void decryptFile(byte[] dek, byte[] nonce, File input, File output) throws Exception {
    Runtime runtime = Runtime.getRuntime();
    long available = runtime.maxMemory() - (runtime.totalMemory() - runtime.freeMemory());
    decryptFile(dek, nonce, input, output, available);
  }

  static void requireMemoryBudget(long encryptedBytes, long availableBytes) {
    long budget =
      availableBytes <= RESERVE_BYTES
        ? 0
        : Math.min(MAX_ENCRYPTED_BYTES, (availableBytes - RESERVE_BYTES) / 4);
    if (encryptedBytes > budget) throw new IllegalStateException(
      "insufficient_memory_for_encrypted_bundle; encryptedBytes=" +
        encryptedBytes +
        "; budgetBytes=" +
        budget
    );
  }

  static void decryptFile(byte[] dek, byte[] nonce, File input, File output, long availableBytes)
    throws Exception {
    long size = input.length();
    if (size <= TAG_LENGTH) throw new IllegalStateException(
      "invalid bundle encryption parameter: ciphertext too short"
    );
    requireMemoryBudget(size, availableBytes);
    if (dek == null || dek.length != KEY_LENGTH || nonce == null || nonce.length != 12) {
      throw new IllegalStateException("invalid bundle encryption parameter: key or nonce length");
    }
    Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
    cipher.init(
      Cipher.DECRYPT_MODE,
      new SecretKeySpec(dek, "AES"),
      new GCMParameterSpec(GCM_TAG_BITS, nonce)
    );

    // Explicit doFinal (not CipherOutputStream) so a GCM tag failure always
    // throws instead of depending on close() behavior. GCM decryption
    // buffers the full input internally anyway before releasing plaintext.
    byte[] ciphertext = readExactBytes(input, (int) size);
    byte[] plaintext = cipher.doFinal(ciphertext);
    try (FileOutputStream out = new FileOutputStream(output)) {
      out.write(plaintext);
    }
  }

  private static byte[] readExactBytes(File file, int size) throws Exception {
    byte[] bytes = new byte[size];
    try (DataInputStream input = new DataInputStream(new FileInputStream(file))) {
      input.readFully(bytes);
      if (input.read() != -1) throw new java.io.IOException("Encrypted bundle changed during read");
    }
    return bytes;
  }
}
